module/zfs/zvol.c

```diff
@@ -966,13 +968,22 @@ zvol_request(struct request_queue *q, struct bio *bio)
 		 */
 		zvr->rl = zfs_range_lock(&zv->zv_range_lock, offset, size,
 		    RL_WRITER);
+		/*
+		 * FIXME: if there is at least another lock held on the same
+		 * block process this request synchronously.
+		 */
+		if (zfs_range_locked(&zv->zv_range_lock, P2ALIGN(offset,
+		    zv->zv_volblocksize), zv->zv_volblocksize) > 1)
+			need_sync = B_TRUE;
```

Comment:
Exactly right.  Which is what will happen if we replace your `zfs_range_locked()` check  with the same check `sync` check from `zvol_write()`.    All of `zvol_write()` will run synchronously in `zvol_request()` like your existing patch.